### PR TITLE
dc/graphql search

### PR DESCRIPTION
**Pull request by dajinchu**
## Overview
This PR adds search to graphql. You can preview using the playground in browser at `/graphql`

**Note that the return value is a little different from REST.**
We used to send back a `type` field in each result item. GraphQL convention calls it `__typename`. Potato potato, same thing.
Also note that REST search items looked like 
```json
{
  "type": "class",
  "class": {},
  "sections": {}
}
```
Whereas the graphql structure is 
```json
{
   "__typename": "ClassOccurrence",
  "name": "Fundies",
  "subject": "CS",
  "sections": {}
}
```
with the section held as just a property of the class occurrence. I think this just makes more sense logically. ClassOccurences should have references to their sections. Simple.

## Deploying

This PR is low risk. Just adds a graphql query. Does not impact existing functionality.

## Example Query 
```graphql
{
  search(termId: 202130, 
    subject:"CS", 
    classIdRange:{max:5000,min:4000},  
    first: 2){
    totalCount
    pageInfo {
      hasNextPage
    }
    nodes {
      __typename
      ... on Employee {
        name
      }
      ... on ClassOccurrence {
        name
        sections {
          profs
          crn
        }
      }
    }
  }
}
```

**Result**
```json
{
  "data": {
    "search": {
      "totalCount": 15,
      "pageInfo": {
        "hasNextPage": true
      },
      "nodes": [
        {
          "__typename": "ClassOccurrence",
          "name": "Artificial Intelligence",
          "sections": [
            {
              "profs": [
                "Lok Sang Wong"
              ],
              "crn": "30209"
            },
            {
              "profs": [
                "Stacy Marsella"
              ],
              "crn": "38417"
            }
          ]
        },
        {
          "__typename": "ClassOccurrence",
          "name": "Natural Language Processing",
          "sections": [
            {
              "profs": [
                "Felix Muzny"
              ],
              "crn": "32493"
            }
          ]
        }
      ]
    }
  }
}
```

### CODE REVIEW

File: backend/graphql/index.js [hunk 1]
```diff
@@ -25,11 +29,24 @@ const baseQuery = gql`
   type Query {
     _empty: String
   }
+
+  type PageInfo {
```
Comment (gamburgm) [inline backend/graphql/index.js:33]: Don't we only have this property on the search endpoint (and if so, shouldn't this just be a property there)?

File: backend/graphql/resolvers/search.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import { identity, pickBy } from 'lodash';
+import searcher from '../../searcher';
+import { Course, Employee } from '../../types';
+
+type SearchResultItem = Course | Employee;
```
Comment (gamburgm) [inline backend/graphql/resolvers/search.ts:5]: Nit: `SearchResult` vs. `SearchResultItem`, `SearchResultMeta` vs. `SearchResultItemConnection`? I care less about the former than the latter, I don't think `connection` accurately describes the data here

File: backend/graphql/resolvers/search.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  campus?: string[];
+  classType?: string[];
+  classIdRange?: { min: number; max: number };
+
+  offset: number;
+  first: number;
+}
+const resolvers = {
+  Query: {
+    search: async (
```
Comment (gamburgm) [inline backend/graphql/resolvers/search.ts]: Nit but can you put the function declaration on one line?

File: backend/graphql/resolvers/search.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  termId: number;
+  query?: string;
+  subject?: string[];
+  nupath?: string[];
+  campus?: string[];
+  classType?: string[];
+  classIdRange?: { min: number; max: number };
+
+  offset: number;
+  first: number;
```
Comment (gamburgm) [inline backend/graphql/resolvers/search.ts:25]: What is `first` supposed to mean? Is this supposed to be like `first ten` kinda thing? If so, then I think `count` or `size` would be better

File: backend/graphql/resolvers/search.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+const resolvers = {
+  Query: {
+    search: async (
+      parent,
+      args: SearchArgs,
+    ): Promise<SearchResultItemConnection> => {
+      const { offset = 0, first = 10 } = args;
+      const results = await searcher.search(
```
Comment (gamburgm) [inline backend/graphql/resolvers/search.ts:31]: In theory you could destructure here for `resultCount` and `searchContent`, but no big deal either way.

File: backend/graphql/typeDefs/classOccurrence.js [hunk 6]
```diff
@@ -7,11 +7,35 @@ const typeDef = gql`
...
 
-    desc: String
+    desc: String!
     prereqs: JSON
     coreqs: JSON
+    prereqsFor: JSON
+    optPrereqsFor: JSON
     maxCredits: Int
     minCredits: Int
+    classAttributes: [String!]!
```
Comment (gamburgm) [inline backend/graphql/typeDefs/classOccurrence.js:17]: I think it's possible for `classAttributes` to be empty, how would you express that?

File: backend/search_types.ts [hunk 7]
```diff
@@ -208,21 +206,14 @@ export interface AggResults {
...
-export interface SearchResult {
-  [resultId: string]: ResultDoc;
-}
-
-export interface CourseDoc {
-  class: Course;
-  sections: Section[];
-}
-export type ResultDoc = Employee | CourseDoc;
+export type SearchResult =
```
Comment (gamburgm) [inline backend/search_types.ts:214]: This is not directly related to this ticket but should we change the model to not have these `type` declarations?

**Review by gamburgm — CHANGES_REQUESTED** [review]
Looks pretty good but added some comments. Also, the type errors in `updater.test.seq.ts` I think are caused by this PR (we'll get rid of the `dumpProcessor` type errors in the re-write). 

What's the plan for typescript GraphQL?

File: backend/graphql/index.js (hunk 1 above)
Comment (dajinchu) [reply backend/graphql/index.js:33]: It's just a GraphQL convention for pagination. We will likely need it in many places.

File: backend/graphql/resolvers/search.ts (hunk 4 above)
Comment (dajinchu) [reply backend/graphql/resolvers/search.ts:25]: https://graphql.org/learn/pagination/#pagination-and-edges Just GraphQL convention.

File: backend/graphql/resolvers/search.ts (hunk 2 above)
Comment (dajinchu) [reply backend/graphql/resolvers/search.ts:5]: GraphQL convention uses Item and Connection. Some graphql clients like Relay require it to be in this format to auto-handle pagination. 

For example: https://docs.github.com/en/free-pro-team@latest/graphql/reference/queries#searchresultitemconnection

File: backend/graphql/typeDefs/classOccurrence.js (hunk 6 above)
Comment (dajinchu) [reply backend/graphql/typeDefs/classOccurrence.js:17]: Empty is allowed by this type def. `[Type]!` means it must be an array and not null. `[Type!]` means the items must be of Type and not null.
`[]` is an array with items that are not null.

**Review by NEUDitao — APPROVED** [review]

**Review by gamburgm — APPROVED** [review]
:+1:

**Pull request merged**